volatility/framework/plugins/mac/pslist.py

```diff
@@ -85,13 +121,120 @@ def list_tasks(cls,
             else:
                 seen[proc.vol.offset] = 1
 
-            if not filter_func(proc) and kernel_as.is_valid(proc.vol.offset, proc.vol.size):
+            if not filter_func(proc) and kernel_layer.is_valid(proc.vol.offset, proc.vol.size):
```

Comment:
I also noticed some inconsistency in these checks. 

1) I think kernel_layer.is_valid call should switch to is_readable() like we discussed recently and is done in one method.

2) It also seems like is_readable() should be called before filter_func as filter_func accesses the pid member, which would cause an exception if the process is at a bad pointer (which would be caught by is_readable if called first)

Reply:
Sure, happy with both of those, again feel free to change it or leave it unresolved and I'll fix it when I get a chance.  5:)

Reply:
What we have at the moment is `is_valid` on layers, and `is_readable` *only* on pointers.  The `is_readable` method essentially operates on the the target of the object.  We currently don't have an `is_valid` method on objects, because simple objects read all their data on instantiation, but structs do not.  We could add a check that all their data is present before allowing a Struct object to be built, but we don't intentionally, so that if it crosses a memory boundary but we don't need the part of it that isn't accessible, we can still get to the data that's there.  

The general guidance is that we don't check the object wholesale and only use it if it's all there, we check what we need (by catching exceptions).  It essentially ensures we support and promote the ability to process partial structs.  In this instance, I'm ok checking it here because the consumer expects the methods to yield valid procs.  If we wanted to check it, we'd need to update the filter function to pass over those with invalid `pids`, but we'd also still have to check on member accesses later on.  That would become very inconvenient for future consumers of the methods, so I'm ok checking the entire thing as they're being listed.  However, I don't want to make checking of the entirety of an object easier, so I would not add an is_valid method to normal objects.

I have changed the order of the checks though.  5;)